Post terms: Add prefix & suffix

dsas opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Allowing an optional prefix and suffix to be added to the Post Categories and Post Tags

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
A common pattern in themes is to display categories and tags prepended with a text. This is the case for all default themes, including TwentyTwentyOne.

![image](https://user-images.githubusercontent.com/93301/164783132-489a3543-4a7f-4d33-8321-80463fd53df1.png)

Fixes #29909

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
 * Add prefix and suffix attributes to the post-terms block. 
 * Allow these attributes to be set inside the block editor via a RichText component

## Testing Instructions
 1. Open a Post or page
 2. Insert a Post Tags or Post Categories block
 3. Click the placeholder and start typing
 8. Press Publish
 9. Look at the site on the frontend

<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

## Screenshots or screencast <!-- if applicable -->
![image](https://user-images.githubusercontent.com/93301/164788905-b91eb2fd-df6e-45df-a509-5ab694ebbc38.png)


https://user-images.githubusercontent.com/93301/164788672-b7899105-19d2-4481-ad9b-b5f1153b3e8c.mp4

dsas committed:
Post terms: Add prefix & suffix

Add prefix and suffix attributes to the post-terms block. These
attributes are displayed inside the block div, immediately before and
after the list. The attributes can be edited as rich text inside the
block editor.

Fixes #29909

dsas committed:
Update integration test for post-terms

carolinan commented:
Editing the placeholders works well in my test, the toolbar options also work and the text displays as expected on the front.

Are these terms, prefix and suffix, common enough to be understood by most users?

In my _opinion_, to allow theme developers to style the prefix and suffix, the markup in the editor and front needs to match and the class names `wp-block-post-terms__suffix` and `wp-block-post-terms__prefix` should be used both in the editor and the front.
I think that extensibility will be important, even though it means adding two more spans.

dsas committed:
Post terms: Wrap prefix & suffix in spans

If the prefix or suffix attributes are provided, wrap them in spans with
a classname.

See #40559

dsas commented:
@carolinan thanks for the review, I've made the span change as suggested.

I think 'prefix' and 'suffix' are common enough, but as a native English speaking developer, perhaps I'm not the best person to answer! I don't think there is a simple alternative to 'prefix' and 'suffix' that's as recognisable. Perhaps 'preface' rather than prefix, but I'm not sure that's much clearer really. I'll continue to think about it, but if anyone has any suggestions, they're very welcome.

amustaque97 commented on packages/block-library/src/post-terms/index.php:
```diff
@@ -36,12 +36,22 @@ function render_block_core_post_terms( $attributes, $content, $block ) {
 
+	$prefix = "<div $wrapper_attributes>";
+	if ( isset( $attributes['prefix'] ) && $attributes['prefix'] ) {
+		$prefix .= '<span class="wp-block-post-terms__prefix">' . $attributes['prefix'] . '</span>';
```
Thank you for working on this. Tested the functionality and I can see prefix and suffix working. I'm afraid we should escape the content of both `prefix` and `suffix` before passing it inside `get_the_term_list `.  Let me know what do you think?

dsas replied:
I don't think it should be escaped, but I could very well be wrong. 

Adding some HTML i.e. that which `InnerText` has toolbar buttons for, is a feature, and stripping it would lead to a weird editing experience - what you see in the editor is very different to what you get when viewing the site.

Additionally I'm not aware of what security problem it poses over just using e.g. the paragraph block and the code editor view to add:
```
<!-- wp:paragraph -->
<p><script>alert('hello');</script></p>
<!-- /wp:paragraph -->
```

So far as I know WordPress is reliant on `wp_kses_post` when saving to validate the content is ok, and that might strip javascript from the submitted content, depending upon the type of account you have on that site.


That said, stripping all HTML is the approach taken by e.g. the post-author block:
https://github.com/WordPress/gutenberg/blob/c5b9ca14e68ad6ee2aefa2973589f36cac388a7e/packages/block-library/src/post-author/index.php#L42

but that easily leads to output like this when viewing the site, so I'm not sure it's a good role model.
![image](https://user-images.githubusercontent.com/93301/165187328-4657e3c4-cc7c-4f6d-8383-6354d89357cf.png)

dsas replied:
Filed #40600 to deal with this in the post author block, either way, it's currently a confusing user experience on that block.

carolinan replied:
I agree that it doesn't need to be escaped here between the span tags.

carolinan commented on packages/block-library/src/post-terms/index.php:
```diff
@@ -36,12 +36,22 @@ function render_block_core_post_terms( $attributes, $content, $block ) {
+
+	$suffix = '</div>';
+	if ( isset( $attributes['suffix'] ) && $attributes['suffix'] ) {
+		$suffix = '<span class="wp-block-post-terms__suffix">' . $attributes['suffix'] . $suffix;
```
Is the closing `</span>` missing here?

dsas replied:
oops, well that's embarrassing, thanks for picking that up @carolinan

dsas committed:
Post terms: Add missing closing span

See #40559

Mamaduka commented:
Hi, @amustaque97

What method are you using to merge merging the PRs? Commits from this branch merged into `trunk` without squashing them into one.

It's not a big deal, I just wanted to check if this was an issue on the GitHub side.

amustaque97 commented:
> Hi, @amustaque97
> 
> What method are you using to merge merging the PRs? Commits from this branch merged into `trunk` without squashing them into one.
> 
> It's not a big deal, I just wanted to check if this was an issue on the GitHub side.

Hi @Mamaduka  :wave: , yeah I choose the other option without squashing them into one. I have noticed this behaviour for others, I can see it working correctly but in my case it somehow makes a single commit while merge. 

Next time, I will record my screen and then probably we can discuss.

Mamaduka commented:
@amustaque97, it's supposed to make a single commit on merge. So I would recommend always using the "Squash and merge" option.

![CleanShot 2022-05-01 at 18 16 55](https://user-images.githubusercontent.com/240569/166149992-b9cf834e-8e6c-4636-a9a2-9f2caa12d30c.png)
